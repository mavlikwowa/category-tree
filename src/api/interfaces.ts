/** Свойства ответа метода getTree */
export interface GetTreeProps {
  /** Успешность ответа */
  success: boolean;
  /** Данные */
  data: CategoryProps[];
}

/** Свойство категории */
export interface CategoryProps {
  id: number;
  /** Отображаемое название */
  name: string;
  /** Мультипликатор цены */
  price_multiplier: number;
  /** id игры, в которой опубликована категория */
  game_id: number;
  /** timestmap создания */
  created_at: number;
  /** timestamp последнего обновления */
  updated_at: number;
  /** Порядок отображения */
  sort: number;
  /** Комиссия сервиса */
  commission: number;
  /** Показывать цену */
  show_price: number;
  /** Показывать количество */
  show_count: number;
  /** Показывать время */
  show_time: number;
  /** Можно продавать несколько раз */
  multiple_sale: number;
  /** Текущий статус: 0 - не видна, 1 - видна всем */
  status: number;
  /** ID родителя или null, если категория корневая */
  parent_id: number | null;
  /** Моментальная доставка? */
  is_instant_delivery: number;
  /** Можно создать несколько предложений в этой категории? */
  is_allow_multiple_offers: number | null;
  /** Есть ли вложенные категории: true/false */
  isFinal: boolean;
  /** Дочерние категории, которых может быть неограниченное количество */
  children?: CategoryProps[];
}
