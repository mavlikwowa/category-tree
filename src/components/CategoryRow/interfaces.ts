/** Интерфейсы */
import { CategoryProps } from '../../api/interfaces';
import { SetStateAction, Dispatch } from 'react';

export interface CategoryRowProps extends CategoryProps {
  /** Функция изменения флага */
  setShowChildren?: Dispatch<SetStateAction<boolean>>;
  /** Флаг отображения вложенных категорий */
  showChildren?: boolean;
}
