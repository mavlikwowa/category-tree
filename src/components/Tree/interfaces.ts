/** Интерфейсы */
import { CategoryProps } from '../../api/interfaces';
/** Actions */
import { changeTree } from '../../actionCreators/treeCreator';

/** Интерфейс dispatch'ей компонента */
export interface DispatchProps {
  changeTree: typeof changeTree;
}
/** Интерфейс state store компонента */
export interface StateProps {
  categories: CategoryProps[];
}
