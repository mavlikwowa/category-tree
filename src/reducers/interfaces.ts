import { CategoryProps } from '../api/interfaces';

export interface TreeReduxAction {
  type: TreeActionTypes;
  payload?: CategoryProps[];
}

/** Свойства Action Thunk */
export interface AppThunkAction<TAction> {
  (dispatch: <TAction>(action: TAction) => TAction, getState: () => any): void;
}

/** Список событий для treeReducer */
export enum TreeActionTypes {
  GET_TREE = 'GET_TREE',
}
