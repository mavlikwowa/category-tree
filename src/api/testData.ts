/** Интерфейсы */
import { GetTreeProps } from './interfaces';

/** Заглушка метода получения категорий */
export const TEST_TREE: GetTreeProps = {
  success: true,
  data: [
    {
      id: 6,
      name: 'Золото',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1555732503,
      updated_at: 1555732503,
      sort: 5,
      commission: 3,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: true,
    },
    {
      id: 7,
      name: 'Рейды',
      price_multiplier: 1,
      game_id: 3,
      created_at: 1555816711,
      updated_at: 1555816711,
      sort: 6,
      commission: 10,
      show_price: 0,
      show_count: 0,
      show_time: 0,
      multiple_sale: 0,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: false,
      children: [
        {
          id: 8,
          name: "Дазар'Алор",
          price_multiplier: 1,
          game_id: 3,
          created_at: 1555817073,
          updated_at: 1555817073,
          sort: 1,
          commission: 10,
          show_price: 0,
          show_count: 0,
          show_time: 0,
          multiple_sale: 1,
          status: 1,
          parent_id: 7,
          is_instant_delivery: 0,
          is_allow_multiple_offers: null,
          isFinal: true,
        },
        {
          id: 9,
          name: 'Горнило штормов',
          price_multiplier: 1,
          game_id: 3,
          created_at: 1555919373,
          updated_at: 1555919373,
          sort: 3,
          commission: 10,
          show_price: 0,
          show_count: 1,
          show_time: 1,
          multiple_sale: 1,
          status: 1,
          parent_id: 7,
          is_instant_delivery: 0,
          is_allow_multiple_offers: null,
          isFinal: true,
        },
      ],
    },
    {
      id: 19,
      name: 'Серебро',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1559631163,
      updated_at: 1559631163,
      sort: 13,
      commission: 10,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: 1,
      isFinal: true,
    },
    {
      id: 51,
      name: 'Серебро',
      price_multiplier: 1000,
      game_id: 3,
      created_at: 1573635179,
      updated_at: 1574063841,
      sort: 25,
      commission: 10,
      show_price: 1,
      show_count: 1,
      show_time: 1,
      multiple_sale: 1,
      status: 1,
      parent_id: null,
      is_instant_delivery: 0,
      is_allow_multiple_offers: null,
      isFinal: true,
    },
  ],
};
