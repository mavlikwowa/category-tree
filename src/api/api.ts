/** Библиотеки */
import axios from 'axios';
/** Константы */
import { API_URL } from './config';
import { TEST_TREE } from './testData';
/** Интерфейсы */
import { CategoryProps } from './interfaces';

/**
 * Возвращает дерево категорий
 */
export const getTree = async (): Promise<CategoryProps[] | void> => {
  try {
    // const res = await axios.get(`${API_URL}/categories/tree`);
    // if (!res.data.success) throw 'request error';
    // return res.data.data;
    const res = await new Promise(resolve => setTimeout(() => { resolve(TEST_TREE.data) }, 1000));
    return res as CategoryProps[];
  } catch (e) {
    console.log(e);
  }
};

/**
 * Меняет позицию категории
 * @param categoryId - идентификатор категории
 * @param sort - индекс сортировки
 */
export const changeCategory = async (
  categoryId: number,
  sort: number,
): Promise<void> => {
  try {
    const res = await axios({
      method: 'post',
      url: `${API_URL}/categories/${categoryId}/sort/${sort}`,
      headers: {},
    })
    // if (!res.data.success) throw 'request error';
  } catch (e) {
    console.log(e);
  }
}
