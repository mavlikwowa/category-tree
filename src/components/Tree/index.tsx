/** Библиотеки */
import React, { FC, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import {
  DragDropContext,
  Droppable,
  Draggable,
  DroppableProvided,
  DraggableProvided,
  DropResult,
} from 'react-beautiful-dnd';
/** Компоненты */
import { Table } from 'react-bootstrap';
import CategoryRow from '../CategoryRow';
/** Actions */
import { changeTree } from '../../actionCreators/treeCreator';
/** Интерфейсы */
import { CategoryProps } from '../../api/interfaces'
import { DispatchProps, StateProps } from './interfaces';
/** API */
import { getTree, changeCategory } from '../../api/api'

/** Компонент с деревом категорий */
const Tree: FC<DispatchProps & StateProps> = ({
  categories,
  changeTree
}: DispatchProps & StateProps) => {
  /** Получает данные по категориям */
  const getData = async () => {
    const res = await getTree();
    if (!res) return;
    changeTree(res.sort(sortCategories) as CategoryProps[]);
  };

  /** Флаг показа дочерних категорий */
  const [showChildren, setShowChildren] = useState<boolean>(false);

  useEffect(() => {
    getData();
  }, []);

  /** Формирует новый список категорий */
  const onDragEnd = (result: DropResult) => {
    if (result.combine) {
      categories.splice(result.source.index, 1)
      changeTree(categories);
      return;
    }
    if (!result.destination) {
      return;
    }
    const startIndex = result.source.index;
    const endIndex = result.destination.index;
    const newSortPosition = startIndex - endIndex;

    // В данном случае ждать ответа не обязательно
    changeCategory(categories[startIndex].id, newSortPosition);

    const [removed] = categories.splice(startIndex, 1);
    categories.splice(endIndex, 0, removed);

    changeTree(categories);
  };

  /** Сортирует категории по индексу сортировки */
  const sortCategories = (prev: CategoryProps, next: CategoryProps) => {
    if (prev.sort > next.sort) return 1;
    if (prev.sort < next.sort) return -1;
    return 0;
  };

  return (
    <Table bordered hover>
      <thead>
      <tr>
        <th>Наименование</th>
        <th>Цена</th>
        <th>Идентификатор игры</th>
        <th>Дата создания</th>
        <th>Дата обновления</th>
        <th>Комиссия сервиса</th>
        <th>Показывать цену</th>
        <th>Показывать количество</th>
        <th>Показывать время</th>
        <th>Можно продавать несколько раз</th>
        <th>Статус</th>
        <th>Моментальная доставка</th>
        <th>Создание нескольких предложений в категории</th>
        <th />
      </tr>
      </thead>
      <DragDropContext onDragEnd={onDragEnd} onDragStart={e => setShowChildren(false)}>
        <Droppable droppableId="droppable" isCombineEnabled>
          {(provided: DroppableProvided): JSX.Element => (
            <tbody ref={provided.innerRef}>
              {categories?.length && categories.map((item: CategoryProps, index: number) => {
                return (
                  <React.Fragment key={item.id}>
                    <Draggable key={item.id} draggableId={String(item.id)} index={index}>
                      {(provided: DraggableProvided): JSX.Element => (
                        <tr
                          ref={provided.innerRef}
                          {...provided.draggableProps}
                          {...provided.dragHandleProps}
                          className={`${showChildren && !item.isFinal && 'row_opened'}`}
                        >
                          <CategoryRow
                            showChildren={showChildren}
                            setShowChildren={setShowChildren}
                            {...item}
                          />
                        </tr>
                      )}
                    </Draggable>
                    {showChildren && item.children?.length && item.children.map(child => {
                        return (
                          <tr key={child.id} className={`${showChildren && 'row_child'}`}>
                            <CategoryRow {...child} />
                          </tr>
                        )
                    })}
                  </React.Fragment>
                )
              })}
            </tbody>
          )}
        </Droppable>
      </DragDropContext>
    </Table>
  );
}

const mapStateToProps = (state: CategoryProps[]): StateProps  => ({
  categories: state,
});
const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => bindActionCreators({
  changeTree,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Tree);
