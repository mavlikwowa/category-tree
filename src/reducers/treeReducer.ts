/** Библиотеки */
import { Reducer } from 'redux';

/** Интерфейсы */
import { CategoryProps } from '../api/interfaces';
import { TreeReduxAction, TreeActionTypes } from './interfaces';

const initialState = {};
/**
 * Reducer для получения дерева категорий
 * @param state
 * @param action
 */
export const treeReducer: Reducer = (
  state: CategoryProps[] | {} = initialState,
  action: TreeReduxAction
): CategoryProps[] | {} => {
  switch (action.type) {
    case TreeActionTypes.GET_TREE:
      return action.payload || state;
  }
  return state;
}
