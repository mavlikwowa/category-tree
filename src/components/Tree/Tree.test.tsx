/* Библиотеки */
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { treeReducer } from '../../reducers/treeReducer';

/* Компоненты */
import Tree from './index';
/* Интерфейсы */
import { StateProps, DispatchProps } from './interfaces';
/* Тестовые данные */
import { TEST_TREE } from '../../api/testData';

export function createTestStore() {
  const store = createStore(treeReducer);
  return store;
}

/** Настройки */
const setUp = () => {
  /* mock для функции changeTree */
  const mockChangeTree = jest.fn();
  /* Тестовые данные */
  const props: DispatchProps & StateProps = {
    categories: TEST_TREE.data,
    changeTree: mockChangeTree,
  };
  /* Компонент */
  const wrapper = shallow(
    <Provider store={createTestStore()}>
      <Tree {...props} />
    </Provider>
  );

  return { wrapper, props };
};

describe('Test component Link', () => {
  let wrapper: ShallowWrapper;
  let props: DispatchProps & StateProps;

  beforeEach(() => {
    ({ wrapper, props } = setUp());
  });

  afterEach(() => {
    if (!wrapper) return;
    wrapper.unmount();
  });

  it('should render component', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
