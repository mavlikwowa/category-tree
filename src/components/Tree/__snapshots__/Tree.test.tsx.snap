// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Link should render component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Tree)
    categories={
      Array [
        Object {
          "commission": 3,
          "created_at": 1555732503,
          "game_id": 3,
          "id": 6,
          "isFinal": true,
          "is_allow_multiple_offers": null,
          "is_instant_delivery": 0,
          "multiple_sale": 1,
          "name": "Золото",
          "parent_id": null,
          "price_multiplier": 1000,
          "show_count": 1,
          "show_price": 1,
          "show_time": 1,
          "sort": 5,
          "status": 1,
          "updated_at": 1555732503,
        },
        Object {
          "children": Array [
            Object {
              "commission": 10,
              "created_at": 1555817073,
              "game_id": 3,
              "id": 8,
              "isFinal": true,
              "is_allow_multiple_offers": null,
              "is_instant_delivery": 0,
              "multiple_sale": 1,
              "name": "Дазар'Алор",
              "parent_id": 7,
              "price_multiplier": 1,
              "show_count": 0,
              "show_price": 0,
              "show_time": 0,
              "sort": 1,
              "status": 1,
              "updated_at": 1555817073,
            },
            Object {
              "commission": 10,
              "created_at": 1555919373,
              "game_id": 3,
              "id": 9,
              "isFinal": true,
              "is_allow_multiple_offers": null,
              "is_instant_delivery": 0,
              "multiple_sale": 1,
              "name": "Горнило штормов",
              "parent_id": 7,
              "price_multiplier": 1,
              "show_count": 1,
              "show_price": 0,
              "show_time": 1,
              "sort": 3,
              "status": 1,
              "updated_at": 1555919373,
            },
          ],
          "commission": 10,
          "created_at": 1555816711,
          "game_id": 3,
          "id": 7,
          "isFinal": false,
          "is_allow_multiple_offers": null,
          "is_instant_delivery": 0,
          "multiple_sale": 0,
          "name": "Рейды",
          "parent_id": null,
          "price_multiplier": 1,
          "show_count": 0,
          "show_price": 0,
          "show_time": 0,
          "sort": 6,
          "status": 1,
          "updated_at": 1555816711,
        },
        Object {
          "commission": 10,
          "created_at": 1559631163,
          "game_id": 3,
          "id": 19,
          "isFinal": true,
          "is_allow_multiple_offers": 1,
          "is_instant_delivery": 0,
          "multiple_sale": 1,
          "name": "Серебро",
          "parent_id": null,
          "price_multiplier": 1000,
          "show_count": 1,
          "show_price": 1,
          "show_time": 1,
          "sort": 13,
          "status": 1,
          "updated_at": 1559631163,
        },
        Object {
          "commission": 10,
          "created_at": 1573635179,
          "game_id": 3,
          "id": 51,
          "isFinal": true,
          "is_allow_multiple_offers": null,
          "is_instant_delivery": 0,
          "multiple_sale": 1,
          "name": "Серебро",
          "parent_id": null,
          "price_multiplier": 1000,
          "show_count": 1,
          "show_price": 1,
          "show_time": 1,
          "sort": 25,
          "status": 1,
          "updated_at": 1574063841,
        },
      ]
    }
    changeTree={[MockFunction]}
  />
</ContextProvider>
`;
